File: leetcode/1044(最长重复子串)/1044(最长重复子串).cpp
typedef unsigned long long ull;

class Solution {
public:
    int check(string& s, int len){
        ull p = 31;
        ull hash = 0;
        ull power = 1;

        for(int i=0; i<len; ++i) {
            hash = hash * p + (s[i]-'a');
            power *= p;
        }

        unordered_set<ull> seen{hash};
        for(int i=len; i<s.size(); ++i){
            hash = hash * p - power*(s[i-len]-'a') + (s[i]-'a');
            if(seen.count(hash)){
                return i-len+1;
            }
            seen.insert(hash);
        }
        return -1;
    }


    string longestDupSubstring(string s) {
        int l = 0;
        int r = s.size() - 1;
        int mid, res;
        int ans = -1;
        int len = 0;

        while(l <= r){
            mid = l + ((r-l)>>1);
            res = check(s, mid);
            if(res != -1){
                ans = res;
                len = mid;
                l = mid + 1;
            }
            else{
                r = mid - 1;
            }
        }
        if(ans == -1) return "";
        else return s.substr(ans, len);
    }
};

File: leetcode/1044(最长重复子串)/1044(最长重复子串).go
const (
    p int = 31
)

func longestDupSubstring(s string) string {
    n := len(s)
    l := 1
    r := n

    pre := make([]int, n+1)
    pow := make([]int, n+1)
    pre[0] = 0 
    pow[0] = 1 
    for i:=1; i<=n; i++ {
        pre[i] = pre[i-1] * p + int(s[i-1] - 'a')
        pow[i] = pow[i-1] * p
    }

    end := -1
    for l <=r {
        mid := (l + r) / 2

        ok := false 
        vis := make(map[int]struct{})
         
        x := pre[mid]
        vis[x] = struct{}{}
        for i:=mid; i<n; i++ {
            x = x * p - int(s[i-mid] - 'a') * pow[mid] + int(s[i] - 'a')

            _, ok = vis[x]; if ok {
                end = i + 1
                break 
            } else {
                vis[x] = struct{}{}
            }
        }

        if ok {
            l = mid + 1 
        } else {
            r = mid - 1 
        }
    }
    if end == -1 {
        return "" 
    } else {
        return s[end-r:end]
    }
}